### PR TITLE
Reenable Components E2E

### PR DESCRIPTION
**Pull request by HaoK**

### CODE REVIEW

**Comment by HaoK** [comment]
Looks looks like we are blocked until they push node.js to the helix queues (which they are waiting for after preview 4 stabailizes to do)...  these are failing with missing npm errors now 

```
Stack Trace:
   at Microsoft.AspNetCore.E2ETesting.SeleniumStandaloneServer.InitializeInstance(ITestOutputHelper output)
   at Microsoft.AspNetCore.E2ETesting.SeleniumStandaloneServer.GetInstanceAsync(ITestOutputHelper output) in /_/src/Shared/E2ETesting/SeleniumStandaloneServer.cs:line 77
   at Microsoft.AspNetCore.E2ETesting.BrowserFixture.CreateBrowserAsync(String context, ITestOutputHelper output) in /_/src/Shared/E2ETesting/BrowserFixture.cs:line 95
   at Microsoft.AspNetCore.E2ETesting.BrowserTestBase.InitializeAsync() in /_/src/Shared/E2ETesting/BrowserTestBase.cs:line 42
[xUnit.net 00:02:03.16]     Microsoft.AspNetCore.Components.E2ETest.Tests.RoutingTest.CanFollowLinkToOtherPageWithQueryString [FAIL]
[xUnit.net 00:02:03.16]       System.InvalidOperationException : Failed to launch the server.
[xUnit.net 00:02:03.16]       ExitCode: 1
[xUnit.net 00:02:03.16]       Captured output lines:
[xUnit.net 00:02:03.16]       
[xUnit.net 00:02:03.16]       'npm' is not recognized as an internal or external command,
[xUnit.net 00:02:03.17]       operable program or batch file.
[xUnit.net 00:02:03.17]       .
[xUnit.net 00:02:03.17]       Stack Trace:
[xUnit.net 00:02:03.17]            at Microsoft.AspNetCore.E2ETesting.SeleniumStandaloneServer.InitializeInstance(ITestOutputHelper output)
[xUnit.net 00:02:03.17]         /_/src/Shared/E2ETesting/SeleniumStandaloneServer.cs(77,0): at Microsoft.AspNetCore.E2ETesting.SeleniumStandaloneServer.GetInstanceAsync(ITestOutputHelper output)
[xUnit.net 00:02:03.17]         /_/src/Shared/E2ETesting/BrowserFixture.cs(95,0): at Microsoft.AspNetCore.E2ETesting.BrowserFixture.CreateBrowserAsync(String context, ITestOutputHelper output)
[xUnit.net 00:02:03.17]         /_/src/Shared/E2ETesting/BrowserTestBase.cs(42,0): at Microsoft.AspNetCore.E2ETesting.BrowserTestBase.InitializeAsync()
[xUnit.net 00:02:03.18]     Microsoft.AspNetCore.Components.E2ETest.Tests.RoutingTest.ClickingAnchorWithNoHrefShouldNotNavigate [FAIL]
[xUnit.net 00:02:03.18]       System.InvalidOperationException : Failed to launch the server.
[xUnit.net 00:02:03.18]       ExitCode: 1
[xUnit.net 00:02:03.18]       Captured output lines:
[xUnit.net 00:02:03.18]       
[xUnit.net 00:02:03.18]       'npm' is not recognized as an internal or external command,
[xUnit.net 00:02:03.18]       operable program or batch file.
```

**Comment by HaoK** [comment]
Getting a little farther now: error is with selenium-standalone not being found now.

Is this something that should get installed as part of the run or something that we need to pull down before running these Component E2E tests?

@javiercn @SteveSandersonMS 

https://mc.dot.net/#/user/aspnetcore/pr~2Faspnet~2Faspnetcore/ci/20190422.4/workItem/Microsoft.AspNetCore.Components.E2ETests-netcoreapp3.0/wilogs

```
[xUnit.net 00:01:01.96]       > microsoft.aspnetcore.components.e2etest@0.0.1 selenium-standalone C:\dotnetbuild\work\c83b73fd-01bd-4958-a429-55c880f0cd0a\Work\7463a409-fa45-476d-9616-7d1cc52f6610\Exec
[xUnit.net 00:01:01.96]       > selenium-standalone "start" "--" "-port" "50466"
[xUnit.net 00:01:01.96]       
[xUnit.net 00:01:01.96]       'selenium-standalone' is not recognized as an internal or external command,
[xUnit.net 00:01:01.96]       operable program or batch file.
[xUnit.net 00:01:01.96]       npm ERR! code ELIFECYCLE
[xUnit.net 00:01:01.96]       npm ERR! errno 1
[xUnit.net 00:01:01.96]       npm ERR! microsoft.aspnetcore.components.e2etest@0.0.1 selenium-standalone: `selenium-standalone "start" "--" "-port" "50466"`
[xUnit.net 00:01:01.96]       npm ERR! Exit status 1
[xUnit.net 00:01:01.96]       npm ERR! 
[xUnit.net 00:01:01.96]       npm ERR! Failed at the microsoft.aspnetcore.components.e2etest@0.0.1 selenium-standalone script.
[xUnit.net 00:01:01.96]       npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[xUnit.net 00:01:01.96]       npm WARN Local package.json exists, but node_modules missing, did you mean to install?
[xUnit.net 00:01:01.96]       
[xUnit.net 00:01:01.96]       npm ERR! A complete log of this run can be found in:
[xUnit.net 00:01:01.96]       npm ERR!     C:\Users\runner\AppData\Roaming\npm-cache\_logs\2019-04-22T10_26_39_530Z-debug.log
[xUnit.net 00:01:01.96]       

```

**Comment by javiercn** [comment]
It’s installed at build time with yarn install

**Comment by HaoK** [comment]
So for the helix machines which aren't building the app, we'll need to add some custom commands to install things?

**Comment by HaoK** [comment]
Is there a simple runtime command we can do that's equivalent to this?

https://github.com/aspnet/AspNetCore/blob/5889aea06cc7ed5c2f743e2fd8ed8abcb7c2d957/src/Middleware/NodeServices/src/Microsoft.AspNetCore.NodeServices.csproj#L24

Like before running the tests, do a 'yarn install packages.json' in the directory?

**Comment by javiercn** [comment]
You need the packages.json along with the yarn.lock and you should have yarn installed globally.

**Comment by HaoK** [comment]
So include packages.json and yarn.lock, and then run "yarn install" before running the tests should do the trick?

File: src/Shared/E2ETesting/SeleniumStandaloneServer.cs [hunk 1]
```diff
@@ -112,7 +112,7 @@ private static async Task InitializeInstance(ITestOutputHelper output)
             {
                 // Just create a random tracking folder on helix
                 trackingFolder = Path.Combine(Directory.GetCurrentDirectory(), Path.GetRandomFileName());
-                Directory.Create(trackingFolder));
+                Directory.CreateDirectory(trackingFolder));
```
Comment (javiercn) [inline src/Shared/E2ETesting/SeleniumStandaloneServer.cs]: What's the difference between these two?

Comment (HaoK) [reply src/Shared/E2ETesting/SeleniumStandaloneServer.cs]: One actually exists :)

**Comment by HaoK** [comment]
@javiercn before I dig too much farther, these tests should run on OSX and linux right?

Currently seeing 
```
 System.ComponentModel.Win32Exception : No such file or directory
Stack Trace:
   at System.Diagnostics.Process.ForkAndExecProcess(String filename, String[] argv, String[] envp, String cwd, Boolean redirectStdin, Boolean redirectStdout, Boolean redirectStderr, Boolean setCredentials, UInt32 userId, UInt32 groupId, UInt32[] groups, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd, Boolean usesTerminal, Boolean throwOnNoExec)
   at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
   at System.Diagnostics.Process.Start(ProcessStartInfo startInfo)
   at Microsoft.AspNetCore.E2ETesting.SeleniumStandaloneServer.InitializeInstance(ITestOutputHelper output) in /_/src/Shared/E2ETesting/SeleniumStandaloneServer.cs:line 130
   at Microsoft.AspNetCore.E2ETesting.SeleniumStandaloneServer.GetInstanceAsync(ITestOutputHelper output) in /_/src/Shared/E2ETesting/SeleniumStandaloneServer.cs:line 77
   at Microsoft.AspNetCore.E2ETesting.BrowserFixture.CreateBrowserAsync(String context, ITestOutputHelper output) in /_/src/Shared/E2ETesting/BrowserFixture.cs:line 95
   at Microsoft.AspNetCore.E2ETesting.BrowserTestBase.InitializeAsync() in /_/src/Shared/E2ETesting/BrowserTestBase.cs:line 42
Failed   Microsoft.AspNetCore.Components.E2ETest.ServerExecutionTests.ServerEventTest.PreventDefault_AppliesToFormOnSubmitHandlers
Error Message:
```

It seems to be weird getting Win32 exceptions on Non Windows machines...?

**Comment by javiercn** [comment]
We only run them on windows, but we have plans to run them on Linux and MAC too. I believe @ryanbrandenburg was working on that.

**Comment by ryanbrandenburg** [comment]
Actually, I found that all of our tests actually [already run on Linux and Mac](https://dnceng.visualstudio.com/internal/_build/results?buildId=165537&view=ms.vss-test-web.build-test-results-tab)? That's weird to me considering it's [specifically disabled](https://github.com/aspnet/AspNetCore/blob/master/src/Shared/E2ETesting/E2ETesting.props#L6).  My preliminary guess is that a property got messed up when we added the RunTemplatesTests property to split the tests into different tasks.

**Comment by javiercn** [comment]
> Actually, I found that all of our tests actually [already run on Linux and Mac](https://dnceng.visualstudio.com/internal/_build/results?buildId=165537&view=ms.vss-test-web.build-test-results-tab). That's weird to me considering it's [specifically disabled](https://github.com/aspnet/AspNetCore/blob/master/src/Shared/E2ETesting/E2ETesting.props#L6). My preliminary guess is that a property got messed up when we added the RunTemplatesTests property to split the tests into different tasks.

No they don't. The Selenium part only runs on Windows. Other parts run cross-plat. (Create/Restore/Build/Publish, etc).

**Comment by HaoK** [comment]
@MattGal looks like something got borked after we install yarn on windows, any ideas what happened to this job?

https://mc.dot.net/#/user/aspnetcore/pr~2Faspnet~2Faspnetcore/ci/20190427.5/workItem/Microsoft.AspNetCore.Components.E2ETests-netcoreapp3.0/wilogs

```
Windows.10.Amd64.ClientRS4.VS2017.Open
Details from Job a096b9ca-762c-429a-902c-764f5c662eba
ExitCode: 0 
Ran on Machine: a000I34 
Console 
 Executed on a000I34

C:\dotnetbuild\work\a096b9ca-762c-429a-902c-764f5c662eba\Work\c24d1c39-8242-4683-a274-4f318e18e6b3\Exec>npm i yarn 
npm notice created a lockfile as package-lock.json. You should commit this file.
+ yarn@1.15.2
added 1 package and audited 1 package in 2.719s
found 0 vulnerabilities
```

The other non windows queues were able to continue after the yarn install.

**Comment by MattGal** [comment]
@HaoK this happened because your command caused exit to happen at that point with code 0.  I would expect that something is wrong with how the script being called is invoked which is promoting some other thing's `exit /b 0` or equivalent into an exit.  I'd first suggest you improve logging around how this is being invoked, then use the local or VM-based repro tools if this doesn't make the problem more obvious.

**Comment by MattGal** [comment]
I tried this Kusto query:
```
WorkItems | where Started > now(-10d) | where FriendlyName == "Microsoft.AspNetCore.Components.E2ETests-netcoreapp3.0" | join kind= inner (
   Jobs 
| where QueueName startswith "Windows"
| order by Started desc | project  QueueName , JobId, Build, Type, Source 
) on JobId  
| order by Finished desc 
```
and I find no evidence that this worked on "some other Windows queue" or was ever run there.  

I suggest trying the guidance here: https://github.com/npm/npm/issues/2938

**Comment by HaoK** [comment]
Thanks @MattGal yeah the other queues were non windows that weren't hitting this issue

**Comment by HaoK** [comment]
Almost there it looks like, @javiercn hopefully the last issue is just another case of resolving the content root properly on helix. 

Its looking for the sources using repo root which isn't there on helix
```
[xUnit.net 00:00:19.36]       System.ArgumentException : The content root 'F:\workspace\_work\1\s\src\Components\test\testassets\BasicTestApp' does not exist.
```

Looks like the logic is coming from here https://github.com/aspnet/AspNetCore/blob/helix-comp/src/Components/test/E2ETest/Infrastructure/ServerFixtures/ServerFixture.cs#L31 which is using the props setting the content root here: https://github.com/aspnet/AspNetCore/blob/5889aea06cc7ed5c2f743e2fd8ed8abcb7c2d957/src/Shared/E2ETesting/E2ETesting.targets#L93

Are you ok if I add a if helix branch which just hardcodes the content root to AppContext.BaseDirectory + "/testassets/AppName" which we can just include everything testassets in the helix payload

**Comment by HaoK** [comment]
Replaced by https://github.com/aspnet/AspNetCore/pull/10997

**Pull request closed without merging**